### PR TITLE
Track source mappings through PY->TS

### PR DESCRIPTION
**Pull request by darzu**
Needed for correct error messages, future debugger support, and possibly better intellisense.

Fixes: https://github.com/microsoft/pxt-minecraft/issues/1614
Related: https://github.com/microsoft/pxt-minecraft/issues/1673

TODO:
- [x] in `setDiagnostics`, use the source map to show TS errors in PY

Now:
Screen Shot 2020-01-28 at 6 31 10 PM

Whereas before we saw no error. (Note that the TurnDirection vs SixDirection confusion is still very confusing)

https://minecraft.makecode.com/app/c16ae7fa3eeabd09f4440341c0da58afc2f1a1e1-f4abc06a71

### CODE REVIEW

File: pxtlib/jsoutput.ts [hunk 1]
```diff
@@ -273,7 +273,9 @@ namespace pxt.blocks {
     export interface SourceInterval {
         id: string;
         start: number;
+        startPos: number;
```
Comment (pelikhan) [inline pxtlib/jsoutput.ts]: please document the difference with start.

File: pxtlib/service.ts [hunk 2]
```diff
@@ -98,6 +98,14 @@ namespace ts.pxtc {
         value: number;
     }
 
+    export type CodeLang = "py" | "blocks" | "ts"
+    export type SourceInterval = {
```
Comment (pelikhan) [inline pxtlib/service.ts]: it's confusing to use the same typename in different places

Comment (darzu) [reply pxtlib/service.ts]: ty, renamed for clarification

File: pxtlib/jsoutput.ts (hunk 1 above)
Comment (darzu) [reply pxtlib/jsoutput.ts]: ty, added clarification

**Comment by darzu** [comment]
FYI @riknoll , I think this can help with PY grey blocks

File: pxtcompiler/emitter/transpile.ts [hunk 3]
```diff
@@ -10,15 +10,16 @@ namespace ts.pxtc.transpile {
     export interface TranspileResult {
         diagnostics: pxtc.KsDiagnostic[],
         success: boolean,
-        outfiles: { [key: string]: string }
+        outfiles: { [key: string]: string },
```
Comment (pelikhan) [inline pxtcompiler/emitter/transpile.ts]: pxt.Map<string>

File: pxtlib/jsoutput.ts [hunk 4]
```diff
@@ -270,15 +270,17 @@ namespace pxt.blocks {
         ".": 18,
     }
 
-    export interface SourceInterval {
+    export interface SourceIntervalToId {
         id: string;
-        start: number;
-        end: number;
+        startLine: number; // 0-indexed
```
Comment (pelikhan) [inline pxtlib/jsoutput.ts]: let's add docs about what those mean

File: pxtlib/jsoutput.ts [hunk 5]
```diff
@@ -381,15 +384,21 @@ namespace pxt.blocks {
...
+                    node.startLine = Math.min(node.startLine, start);
+                    node.endLine = Math.max(node.endLine, end);
+                    node.startPos = Math.min(node.startPos, startPos);
+                    node.endPos = Math.max(node.endPos, endPos);
                 }
                 else {
-                    const interval = { id: n.id, start: start, end: end }
+                    const interval: SourceIntervalToId = {
+                        id: n.id,
+                        startLine: start, startPos, endLine: end, endPos
```
Comment (pelikhan) [inline pxtlib/jsoutput.ts]: rename local variables to "startLine" and "endLine

File: pxtlib/service.ts [hunk 6]
```diff
@@ -98,6 +98,18 @@ namespace ts.pxtc {
         value: number;
     }
 
+    export type CodeLang = "py" | "blocks" | "ts"
+    export interface SourceInterval {
+        ts: {
+            start: number;
```
Comment (pelikhan) [inline pxtlib/service.ts]: startLine or startPos?

**Pull request merged**